sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -497,10 +497,32 @@ case class Lower(child: Expression)
 abstract class StringPredicate extends BinaryExpression
   with Predicate with ImplicitCastInputTypes with NullIntolerant {
 
+  final val collationId: Int = left.dataType.asInstanceOf[StringType].collationId
```

Comment:
I think it should be a lazy val, in case we fail with java cast issue before the `checkInputDataTypes` is called.

Reply:
makes sense! fixed